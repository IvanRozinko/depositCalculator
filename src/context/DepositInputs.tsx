import { createContext, useContext } from "react"

export interface IDeposit {
  duration: number,
  percent: number,
  sum: number,
  monthly: number,
  withoutFirstMonth: boolean,
  longation: number,
  percentDestination: string
}

export interface DepositDefaultContext{
  deposit: IDeposit
  setContextValue: (newDeposit: IDeposit) => void
}

export const Deposit: IDeposit = {
  duration: 4,
  percent: 6,
  sum: 100000,
  monthly: 10000,
  withoutFirstMonth: false,
  longation: 0,
  percentDestination: 'deposit'
}

export const DepositContext = createContext<DepositDefaultContext>({
  deposit: Deposit,
  setContextValue: (newDeposit: IDeposit):void => {}
})

export const useDepositContext = () => useContext<DepositDefaultContext>(DepositContext)