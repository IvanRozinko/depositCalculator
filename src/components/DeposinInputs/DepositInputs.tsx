import React from 'react'
import { 
  makeStyles, Typography, Checkbox, FormControlLabel, Box
} from '@material-ui/core'; 
import InputSlider from '../InputSlider/InputSlider';
import RateText from '../RateText/RateText';
import SliderLabel from '../SliderLabel/SliderLabel';
import TextInput from '../TextInput/TextInput';
import PercentDestination from '../PercentDestination/PercentDestination';
import { useDepositContext } from '../../context/DepositInputs';
import InfoIcon from '@material-ui/icons/InfoOutlined';

import { duration, rates, longation } from '../../mocks/depositRanks';
import { stripNoneNumbers } from '../../helpers/helpers';
import { MAX_DEPOSIT_SUM, MAX_LONGATIONS } from '../../config/constants';

const useStyles = makeStyles({
  container: {
    flexGrow: 1,
    height: '100%',
    paddingRight: 20
  },
  marginLeft: {
    marginLeft: 10
  }
});

function DepositInputs() {
  const styles = useStyles();
  const { deposit, setContextValue } = useDepositContext();

  const mergeContextValue = (field: string, value: number | boolean | string) => {
    setContextValue({
      ...deposit,
      [field]: value
    })
  }

  const handleRateChange = (e: any, value: number | number[]) => {
    const duration = typeof value === 'number' ? value : value[0];
    setContextValue({
      ...deposit,
      duration,
      percent: getRateByDuration(duration)
    })
  }

  const handleSumChange = (e: any, value: number | number[]) => {
    const sum = typeof value === 'number' ? value : value[0];
    mergeContextValue('sum', sum);
  }

  const normalizeSum = (sum: number) => {
    let normalized = sum;
    if (sum < 0 || isNaN(sum)) {
      normalized = 0;
    } else if ( sum > MAX_DEPOSIT_SUM) {
      normalized = MAX_DEPOSIT_SUM;
    }
    return normalized;
  }
  
  const handleMonthlyChange = (e: any, value: number | number[]) => {
    const monthly = typeof value === 'number' ? value : value[0];
    mergeContextValue('monthly', monthly);
  } 
  
  const handleSumInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const sum = stripNoneNumbers(event.target.value);
    mergeContextValue('sum', normalizeSum(parseInt(sum)));
  }

  const handleMonthlyInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const monthly = stripNoneNumbers(event.target.value)
    mergeContextValue('monthly', normalizeSum(+monthly));
  }

  const handleCheckBox = (event: React.ChangeEvent<HTMLInputElement>) => {
    mergeContextValue('withoutFirstMonth', event.target.checked)
  }

  const getRateByDuration = (duration: number) => rates[`${duration}`] / 10;

  const handleLongationChange = (e: any, value: number | number[]) => {
    const longation = typeof value === 'number' ? value : value[0];
    mergeContextValue('longation', longation);
  } 

  const handlePercentDestination = (newDestination: string) => {
    mergeContextValue('percentDestination', newDestination)
  }

  return (
    <div className={styles.container}>
      <Typography variant='h6'>Стандарт строковий</Typography>
      <InputSlider 
        value={deposit.duration}
        handleChange={handleRateChange}
        marks={duration}
        rate={deposit.percent}
        children={<RateText rate={`${deposit.percent} %`}/>}
        label={<SliderLabel label={`Ставка (на рік)/ Строк (міс.)`} />}
        valueDisplay="on"
        sliderOptions={{
          'aria-labelledby': "discrete-slider-custom",
          min: duration[0].value,
          max: duration[duration.length - 1].value
        }}
      />
      <InputSlider 
        value={deposit.sum}
        handleChange={handleSumChange}
        label={<SliderLabel label="Сума" />}
        children={<TextInput sum={deposit.sum} handleInputChange={handleSumInputChange} />}
        valueDisplay="off"
        sliderOptions={{
          'aria-labelledby': "discrete-slider-custom",
          min: 0,
          max: MAX_DEPOSIT_SUM
        }}
      />
      <InputSlider 
        value={deposit.monthly}
        handleChange={handleMonthlyChange}
        label={<SliderLabel label="Щомісячне поповнення" />}
        children={<TextInput sum={deposit.monthly} handleInputChange={handleMonthlyInputChange} />}
        valueDisplay="off"
        sliderOptions={{
          'aria-labelledby': "discrete-slider-custom",
          min: 0,
          max: MAX_DEPOSIT_SUM
        }}
      />
      <FormControlLabel
        control={
          <Checkbox
            checked={deposit.withoutFirstMonth}
            onChange={handleCheckBox}
            color="primary"
          />
        }
        label="Без поповнення в перший місяць"
      />
       <InputSlider 
        value={deposit.longation}
        marks={longation}
        handleChange={handleLongationChange}
        label={<SliderLabel label="Кількість продовжень" />}
        children={<RateText rate={`${deposit.longation}`}/>}
        valueDisplay="off"
        sliderOptions={{
          'aria-labelledby': "discrete-slider-custom",
          min: 0,
          max: MAX_LONGATIONS
        }}
      />
      <PercentDestination
        destination={deposit.percentDestination}
        handlePercentDestination={handlePercentDestination} 
      />
      <Box
        display='flex'
        flexDirection='row'
        alignItems='center'
        marginTop={2}
      >
        <InfoIcon
          color='primary'
        />
        <Typography
         className={styles.marginLeft}
        >
          Розрахунок процентів має довідковий характер
        </Typography>
      </Box>
    </div>
  )
}

export default DepositInputs
