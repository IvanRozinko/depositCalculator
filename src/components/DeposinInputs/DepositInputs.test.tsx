import DepositInputs from "./DepositInputs";
import { shallow } from 'enzyme';

describe('<DepositInputs />:', () => {
  const wrapper = shallow(
    <DepositInputs />
  )

  it('should renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  })
})