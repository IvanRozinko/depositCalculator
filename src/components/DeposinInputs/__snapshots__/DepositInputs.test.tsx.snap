// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DepositInputs />: should renders correctly 1`] = `
<div
  className="makeStyles-container-1"
>
  <WithStyles(ForwardRef(Typography))
    variant="h6"
  >
    Стандарт строковий
  </WithStyles(ForwardRef(Typography))>
  <InputSlider
    handleChange={[Function]}
    label={
      <SliderLabel
        label="Ставка (на рік)/ Строк (міс.)"
      />
    }
    marks={
      Array [
        Object {
          "label": "3",
          "value": 3,
        },
        Object {
          "label": "4",
          "value": 4,
        },
        Object {
          "label": "5",
          "value": 5,
        },
        Object {
          "label": "6",
          "value": 6,
        },
        Object {
          "label": "7",
          "value": 7,
        },
        Object {
          "label": "8",
          "value": 8,
        },
        Object {
          "label": "9",
          "value": 9,
        },
        Object {
          "label": "10",
          "value": 10,
        },
        Object {
          "label": "11",
          "value": 11,
        },
        Object {
          "label": "12",
          "value": 12,
        },
      ]
    }
    rate={6}
    sliderOptions={
      Object {
        "aria-labelledby": "discrete-slider-custom",
        "max": 12,
        "min": 3,
      }
    }
    value={4}
    valueDisplay="on"
  >
    <RateText
      rate="6 %"
    />
  </InputSlider>
  <InputSlider
    handleChange={[Function]}
    label={
      <SliderLabel
        label="Сума"
      />
    }
    sliderOptions={
      Object {
        "aria-labelledby": "discrete-slider-custom",
        "max": 1000000,
        "min": 0,
      }
    }
    value={100000}
    valueDisplay="off"
  >
    <TextInput
      handleInputChange={[Function]}
      sum={100000}
    />
  </InputSlider>
  <InputSlider
    handleChange={[Function]}
    label={
      <SliderLabel
        label="Щомісячне поповнення"
      />
    }
    sliderOptions={
      Object {
        "aria-labelledby": "discrete-slider-custom",
        "max": 1000000,
        "min": 0,
      }
    }
    value={10000}
    valueDisplay="off"
  >
    <TextInput
      handleInputChange={[Function]}
      sum={10000}
    />
  </InputSlider>
  <WithStyles(ForwardRef(FormControlLabel))
    control={
      <WithStyles(ForwardRef(Checkbox))
        checked={false}
        color="primary"
        onChange={[Function]}
      />
    }
    label="Без поповнення в перший місяць"
  />
  <InputSlider
    handleChange={[Function]}
    label={
      <SliderLabel
        label="Кількість продовжень"
      />
    }
    marks={
      Array [
        Object {
          "label": "1",
          "value": 1,
        },
        Object {
          "label": "2",
          "value": 2,
        },
        Object {
          "label": "3",
          "value": 3,
        },
        Object {
          "label": "4",
          "value": 4,
        },
        Object {
          "label": "5",
          "value": 5,
        },
      ]
    }
    sliderOptions={
      Object {
        "aria-labelledby": "discrete-slider-custom",
        "max": 5,
        "min": 0,
      }
    }
    value={0}
    valueDisplay="off"
  >
    <RateText
      rate="0"
    />
  </InputSlider>
  <PercentDestination
    destination="deposit"
    handlePercentDestination={[Function]}
  />
  <Styled(MuiBox)
    alignItems="center"
    display="flex"
    flexDirection="row"
    marginTop={2}
  >
    <Memo(ForwardRef(InfoOutlinedIcon))
      color="primary"
    />
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-marginLeft-2"
    >
      Розрахунок процентів має довідковий характер
    </WithStyles(ForwardRef(Typography))>
  </Styled(MuiBox)>
</div>
`;
